networkx/algorithms/distance_measures.py

```diff
@@ -769,3 +770,104 @@ def resistance_distance(G, nodeA, nodeB, weight=None, invert_weight=True):
+    """Returns Kemeny's constant of the given graph.
+
+    *Kemeny's constant* of a graph `G` can be computed by regarding the graph
+    as a Markov chain. Kemeny's constant is then the expected number of time steps
+    to transition from a starting state i to a random destination state
+    sampled from the Markov chain's stationary distribution.
+    Kemeny's constant is independent of the chosen initial state [1]_.
+
+    Kemeny's constant is used as a measure of connectivity of graphs.
+    Large values indicate a poorly connected graph whereas low values
+    correspond to well-connected graphs.
+
+    If weight is not provided, then a weight of 1 is used for all edges.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    weight : string or None, optional (default=None)
+       The edge data key used to compute Kemeny's constant.
+       If None, then each edge has weight 1.
+
+    Returns
+    -------
+    K : float
+        Kemeny's constant of the graph `G`.
+
+    Raises
+    ------
+    NetworkXError
+        If the graph `G` is directed.
+
+    NetworkXError
+        If the graph `G` is not connected.
+
+    NetworkXError
+        If the graph `G` does not contain any nodes.
+
+    NetworkXError
+        If the graph `G` contains edges with negative weights.
+
+    Examples
+    --------
+    >>> G = nx.complete_graph(5)
+    >>> round(nx.kemeny_constant(G), 10)
+    3.2
+
+    Notes
+    -----
+    The implementation is based on equation (3.3) in [2]_. Self-loops are included.
+    Multi-edges are contracted in one edge with weight equal to the sum of the weights.
+
+    References
+    ----------
+    .. [1] Wikipedia
+       "Kemeny's constant."
+       https://en.wikipedia.org/wiki/Kemeny%27s_constant
+    .. [2] Lovász L.
+        Random walks on graphs: A survey.
+        Paul Erdös is Eighty, vol. 2, Bolyai Society,
+        Mathematical Studies, Keszthely, Hungary (1993), pp. 1-46
+    """
+    import numpy as np
+    import scipy as sp
+
+    if len(G) == 0:
+        msg = "Graph G must contain at least one node."
+        raise nx.NetworkXError(msg)
+    elif nx.is_directed(G):
+        msg = "Graph G must be undirected."
+        raise nx.NetworkXError(msg)
+    elif not nx.is_connected(G):
+        msg = "Graph G must be connected."
+        raise nx.NetworkXError(msg)
+    elif nx.is_negatively_weighted(G, weight=weight):
+        msg = "The weights of graph G must be nonnegative."
+        raise nx.NetworkXError(msg)
+
+    # Compute matrix H = D^-1/2 A D^-1/2
+    A = nx.adjacency_matrix(G, weight=weight)
+    n, m = A.shape
+    diags = A.sum(axis=1)
+    with sp.errstate(divide="ignore"):
+        diags_sqrt = 1.0 / np.sqrt(diags)
+    diags_sqrt[np.isinf(diags_sqrt)] = 0
+    DH = sp.sparse.csr_array(sp.sparse.spdiags(diags_sqrt, 0, m, n, format="csr"))
+    H = DH @ (A @ DH)
+
+    # Compute eigenvalues of H
+    eig = sp.linalg.eigvalsh(H.todense())
+    eig = sorted(eig)
```

Comment:
I don't think the eigenvalues need to be sorted (based on the rest of the code).

And, its a shame to have to convert `H.todense()` because that could be a memory issue. But I don't see a way around it with the sparse eigen-solvers offered by Scipy in `sp.sparse.linalg`. They don't allow computation of all the eigenvalues.

Reply:
The eigenvalues must be sorted, because the largest eigenvalue (which equals 1) is not used in the formula for Kemeny's constant. You can also see why: then the formula explodes (and gives +infinity as a result)

For the computation of all eigenvalues, a dense matrix is generally preferred, as far as I know. Anyway, the computation of all eigenvalues is O(N^3) operation, making stuff slow for N >= 4,000.